### PR TITLE
Result Filters: Add 'Length' filter

### PR DESCRIPTION
**Pull request by slook**
+ Added: FilterLength ComboEntryBox for Result Filters bar, default filter in Preferences, history config, etc.
+ Added: Not ! operator and split conditions in check_digit() for Length, Size and Bitrate filters, to enable explicit exclusions.
+ Added: Approximate to within +/- filesize tolerance, 0.1 of a unit each way, if the user doesn't enter the exact bytes.
+ Moved: factorize filesize string interpreter code into separate factorize() function in utils.py.
+ Changed: Accept non-integer value, because if the size is factorized then an exact match was practically impossible.
+ Changed: Text strings in Result Filter Usage popover, added 'Length' section.
+ Changed: "Brittany Spears" -> "Aphex Twin", although perhaps a non-copyrighted artist would be better for the example.

image

_Multiple conditions can be in any order, for example..._
`>7:00|<11:00|!7:30|!7:33|!6:46|=12:45|=260|=1:58:30`
_... shows:_
- all tracks between 7 and 11 minutes long, except for...
- hide tracks that are exactly 7:30 or 7:33 or 6:46 long
- and also show songs that are exactly 12:45 long, exactly 260 seconds, and exactly 1h58m30s long.

_For the Length filter, = "exactly" means to within 1 second (decimals are not supported)._
_For the File Size filter, = "exactly" means to within 0.1 MiB (or within 1 MiB if over 100 MiB), unless raw size is specified._

In future, if we add entries programmatically (on_combo_icon_click() or context menu?) then raw Seconds or Filesize are faster to process and dead accurate, therefore more suitable for storing in the Wishlist for the purpose of excluding unwanted results.

This could help us to provide a mechanism to ignore unwanted Wishlist results (see #1887), because any amount of `!` (not) and `=` (exact) conditions are possible, by stringing together !1|!2|!3|... conditions. This would probably be easier and more efficient than creating a database of ignored Wishlist results.

**_Status: Ready for inclusion into master for further testing (feedback welcomed), aims to close feature request https://github.com/nicotine-plus/nicotine-plus/issues/1727_**

### CODE REVIEW

**Comment by mathiascode** [comment]
Thanks! I'll review and test this as soon as possible, but finishing 3.2.2 is my current priority with the free time I have.

File: pynicotine/gtkgui/search.py [hunk 1]
```diff
@@ -887,6 +902,9 @@ def check_filter(self, row):
         if filters["filtertype"] and not self.check_file_type(filters["filtertype"], row[11]):
             return False
 
+        if filters["filterlength"] and not self.check_digit(filters["filterlength"], row[16].get_uint64(), False):
```
Comment (mathiascode) [inline pynicotine/gtkgui/search.py:905]:
```suggestion
        if filters["filterlength"] and not self.check_digit(filters["filterlength"], row[16].get_uint(), False):
```

File: pynicotine/gtkgui/ui/search.ui [hunk 2]
```diff
@@ -159,14 +159,14 @@
                 <property name="margin-start">12</property>
                 <property name="margin-end">6</property>
                 <child>
-                  <object class="GtkFlowBox">
+                  <object class="GtkFlowBox" id="FiltersFlowBox">
```
Comment (mathiascode) [inline pynicotine/gtkgui/ui/search.ui:162]: No need for these new ids.

Comment (slook) [reply pynicotine/gtkgui/ui/search.ui:162]: Okay I shall remove them. My future intention is to provide some sort of mechanism for showing/hiding the visibility to allow the filters bar to be customized.

File: pynicotine/gtkgui/search.py (hunk 1 above)
Comment (slook) [reply pynicotine/gtkgui/search.py]: Agreed it is adequate to allow for files which are up to 68 years long!

**Comment by mathiascode** [comment]
@hboetes Could you check if the strings in this PR sound fine to you?

File: pynicotine/gtkgui/ui/popovers/searchfilters.ui [hunk 3]
```diff
@@ -108,7 +117,7 @@
                         <property name="xalign">0</property>
                         <property name="margin-start">12</property>
                         <property name="label" translatable="yes">For order-insensitive filtering, as well as filtering several exact phrases, vertical bars can be used to separate phrases and words.
-    Example: Spears|Brittany|My beautiful album|hello</property>
+    Example: Twin|Aphex|Analogue Bubblebath|Digeridoo</property>
```
Comment (mathiascode) [inline pynicotine/gtkgui/ui/popovers/searchfilters.ui:120]: As you mentioned, this should be replaced with a non-copyrighted example before the PR is merged.

**Comment by mathiascode** [comment]
Merging for further testing. Follow-up PRs can be created if necessary.

**Comment by slook** [comment]
@redactedscribe Please can you test the new functionality detailed above https://github.com/nicotine-plus/nicotine-plus/pull/1892#issue-1139921545 in the [latest unstable build](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md) of Nicotine 3.3.0.dev1 and provide any feedback you may have about it. Thanks.

**Comment by slook** [comment]
1. Slightly more useful defaults added in #2238 as also suggested in #2141.
2. Adding support for the 1h30m syntax would be desirable, but there are complexity where spaces might be used in the condition (such as 1h 30m), so I shall not be implementing another syntax definition until orphaned parts of the condition strings are properly handled and/or we have a mechanism of providing feedback about syntax errors and/or a method that is able to interpret sub-parameters that contain spaces.
3. Good point, although for now I think excluding results with no time is better done manually if the user has explicitly entered a !0 condition, because a filter may reduce the results down to nothing but it is not obvious whether or not this is accurate or if it is due to some syntax error.
4. The future intention is to implement an `is_filter_valid` flag or similar that would ideally notfiy exactly which condition has failed, or at least indicate that a failure does exist somewhere within the filter entry. The current implmentation of making the in/out TextEntry boxes background red is slightly hacky, so I would like to implement something more suitable for this purpose.

**Pull request merged**